### PR TITLE
feat(api): harden mutating API idempotency and concurrency

### PR DESCRIPTION
**Pull request by dimileeh**
Automatically opened by AWF workspace `ws_d2e18dbef2b347738f0e1892` (agent: `codex`).

**External task ID**: awf-prd15-idempotent-controls-20260426T175117Z

### Task
Read AGENTS.md and docs/awf_prd_v2.2.md before editing. Work strictly TDD: add failing tests first, confirm the failure when practical, implement, then run validation. Use conventional commits. Do not push; AWF owns push, PR creation, monitoring, and auto-merge. Target branch is codex/awf-post-merge-fixes, so auto_merge must stay true. Preserve existing behavior unless the prompt explicitly tightens the contract. If you touch Alembic migrations, final `uv run --python 3.12 --extra dev alembic heads` must print exactly one head.

PRD gap: all mutating APIs need strong idempotency and optimistic concurrency. Implement the first strict slice for sensitive workspace control APIs.

Behavior:
- `POST /v1/workspaces/{id}/cancel`, `POST /v1/workspaces/{id}/stop`, and `DELETE /v1/workspaces/{id}` must honor `Idempotency-Key`.
- Same key + same normalized action payload returns the original/current operation response instead of creating a duplicate operation or duplicate events.
- Same key + different normalized payload returns a structured `409 IDEMPOTENCY_CONFLICT`.
- Missing key should return a structured client error for these sensitive mutating endpoints; choose the repo's existing error style and document it in tests.
- Support optional `If-Match` workspace version on these endpoints. If present and stale, return a structured version/concurrency error and do not mutate. If omitted, preserve compatibility.
- Repeated cancel/stop/destroy after terminal state must remain safe and idempotent.
- Do not add a migration unless absolutely necessary; `Operation.idempotency_key` already exists.

TDD requirements:
- Add focused API tests for missing key, replay same key, conflicting payload, stale `If-Match`, and no duplicate operation/event creation.
- Add service/repository tests only where useful.
- Keep response shapes compatible with the existing console/CLI where possible.

---
Validation: 6 profile command(s) passed inside the workspace container.

### CODE REVIEW

File: src/awf/db/repositories.py [hunk 1]
```diff
@@ -1490,14 +1491,15 @@ def _sync_candidate_readiness(
...
-    elif stale_reason is None and candidate.stale_reason in ("validation_insufficient_tier",):
-        candidate.stale = False
-        candidate.stale_reason = None
+    if recompute_stale:
+        stale_reason, _ = compute_stale_reason(workspace)
+        # If there's an active stale reason, update it. If the reason clears, clear it.
+        if stale_reason is not None:
+            candidate.stale = True
+            candidate.stale_reason = stale_reason
+        elif stale_reason is None and candidate.stale_reason in ("validation_insufficient_tier",):
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/db/repositories.py:1681]: fixed in commit 6f0e151be7b1a914e2308843a001186de15bbc1b

File: src/awf/service/staleness.py [hunk 2]
```diff
@@ -415,9 +415,8 @@ async def _mark_candidate_stale(
         *,
         stale: bool,
     ) -> None:
-        if candidate.stale == stale:
-            return
         candidate.stale = stale
+        candidate.stale_reason = "stale" if stale else None
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/staleness.py]: fixed in commit 36d36ae9c5920adbc14eaa1b7d94ac0334eec5b7

File: src/awf/service/controls.py [hunk 3]
```diff
@@ -298,6 +381,48 @@ async def _require_workspace(
...
+            existing = await operations.get_by_idempotency_key(idempotency_key)
+            if existing is not None:
+                if (
+                    existing.workspace_id != workspace_id
+                    or existing.type != operation_type.value
+                    or existing.payload != payload
+                ):
+                    raise IdempotencyConflictError()
+                workspace = await self._require_workspace(repo, workspace_id)
+                return workspace, existing
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/controls.py:419]: fixed in commit 3798416e01ff74a4f2c606803caa38a4930c406b

File: tests/unit/api/test_workspace_controls_idempotency.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+
+class FakeCleaner:
+    calls: list[dict[str, object]] = []
+
+    async def cleanup(
+        self,
+        *,
+        workspace_id: str,
+        repo_url: str,
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply tests/unit/api/test_workspace_controls_idempotency.py]: fixed in commit b020f4677042f7b32d80f88c6f2209a28d29b37b

File: src/awf/db/repositories.py [hunk 5]
```diff
@@ -1690,6 +1709,15 @@ async def create(
...
         return await self._session.get(Operation, operation_id)
 
+    async def get_by_idempotency_key(self, key: str) -> Operation | None:
+        stmt = (
+            select(Operation)
+            .where(Operation.idempotency_key == key)
+            .order_by(Operation.created_at.asc(), Operation.id.asc())
+            .limit(1)
+        )
+        return (await self._session.execute(stmt)).scalar_one_or_none()
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/db/repositories.py:1903]: fixed in commit 1989921ce5935ba411c70c2b23c6ad0184c27c17
